File: rust/conformance/src/bin/encode_keyframe_delta.rs
// Copyright 2026 Avala AI
// SPDX-License-Identifier: Apache-2.0

//! Conformance runner: encode a `keyframe-delta` file.
//!
//! There is no committed keyframe-delta corpus variant — a separate milestone owns the
//! corpus — so this synthesizes a deterministic sample sequence in memory and writes it
//! with the Rust keyframe-delta writer. The sequence is built to exercise the model rather
//! than merely to be legal: gaussians move (so every non-keyframe sample carries updates),
//! a group of ids is born partway through, another dies later, ids are rotated within each
//! sample (so a decoder that found gaussians by row instead of by identity would be wrong),
//! and the cadence lays down more than one keyframe with chained deltas between them.
//!
//! The harness step around this then makes the claim that matters: the file this encoder
//! produced is read to an identical canonical `states` summary by the Rust decoder and by
//! the Python reference decoder — the only definition of a conforming file that means
//! anything.
//!
//! Usage: encode_keyframe_delta <out.4dgs> [keyframe|chained|cadence-one]

use std::process::ExitCode;

use fourdgs::keyframe_delta_file::{write_sequence, KeyframeDeltaOptions, Sample};
use fourdgs::model::GaussianSet;
use fourdgs::records::{DELTA_MODE_CHAINED, DELTA_MODE_KEYFRAME};

const DURATION: f64 = 2.0;
const SAMPLES: usize = 17;

/// The cadence and delta mode a shape name asks for.
///
/// `cadence-one` is the §11.11 shape — every chunk a keyframe, no delta chunk anywhere,
/// which is what subsumes the reserved `frame-sequence` name. It is here because the binding
/// runners write it and their gates compare against this writer byte for byte; without it
/// that comparison silently skipped a shape while the gate still reported it as checked. The
/// delta mode is left at the default for it, because with no delta in the file there is
/// nothing for it to name — and a binding is free to leave it alone too.
fn shape_of(name: &str) -> Option<(usize, u8)> {
    match name {
        "chained" => Some((8, DELTA_MODE_CHAINED)),
        "keyframe" => Some((8, DELTA_MODE_KEYFRAME)),
        "cadence-one" => Some((1, DELTA_MODE_CHAINED)),
        _ => None,
    }
}

fn main() -> ExitCode {
    let args: Vec<String> = std::env::args().collect();
    if args.len() < 2 || args.len() > 3 {
        eprintln!("usage: encode_keyframe_delta <out.4dgs> [keyframe|chained|cadence-one]");
        return ExitCode::from(2);
    }
    let shape = args.get(2).map(String::as_str).unwrap_or("chained");
    let Some((keyframe_every, delta_mode)) = shape_of(shape) else {
        eprintln!("unknown shape {shape:?}; expected 'keyframe', 'chained' or 'cadence-one'");
        return ExitCode::from(2);
    };
    match run(&args[1], keyframe_every, delta_mode) {
        Ok(note) => {
            println!("{note}");
            ExitCode::SUCCESS
        }
        Err(message) => {
            eprintln!("{message}");
            ExitCode::FAILURE
        }
    }
}

fn run(output: &str, keyframe_every: usize, delta_mode: u8) -> Result<String, String> {
    let samples = synthesize();
    let kd = KeyframeDeltaOptions {
        keyframe_every,
        delta_mode,
        ..Default::default()
    };
    let first = write_sequence(&samples, DURATION, &kd)
        .map_err(|e| format!("{output}: encode failed: {e}"))?;
    let second = write_sequence(&samples, DURATION, &kd)
        .map_err(|e| format!("{output}: the second encode failed: {e}"))?;
    if first != second {
        return Err(format!(
            "{output}: two encodes of one sequence differ; the writer is not deterministic"
        ));
    }

    // The file has to survive both of this implementation's read paths, and they must reach
    // the same populations, before it is worth asking another implementation about it.
    let streamed = fourdgs::keyframe_delta_file::decode_streamed(&first)
        .map_err(|e| format!("{output}: the writer wrote a file it cannot stream-decode: {e}"))?;
    let (indexed, _) = fourdgs::keyframe_delta_file::decode_indexed(&first)
        .map_err(|e| format!("{output}: the writer wrote a file it cannot index-decode: {e}"))?;
    let a = fourdgs_conformance::keyframe_delta_states_json(&streamed);
    let b = fourdgs_conformance::keyframe_delta_states_json(&indexed);
    if a != b {
        return Err(format!(
            "{output}: the two read paths disagree on a file the writer produced"
        ));
    }

    std::fs::write(output, &first).map_err(|e| format!("{output}: {e}"))?;
    Ok(format!(
        "{} samples, {} chunks, {} bytes, deterministic, both read paths agree",
        samples.len(),
        streamed.chunks.len(),
        first.len()
    ))
}

/// A deterministic sample sequence. Reproducible from an LCG so the encoder has no hidden
/// state; the population is chosen to make every group non-empty somewhere in the file.
fn synthesize() -> Vec<Sample> {
    let mut samples = Vec::with_capacity(SAMPLES);
    for i in 0..SAMPLES {
        // t0s tile [0, duration): the last sample starts below the end, and its interval
        // closes at `duration`, so no chunk gets a zero-width span.
        let frac = i as f64 / SAMPLES as f64;
        let t0 = frac * DURATION;

        // Which ids are live at this sample. The core is always present; one band is born
        // partway in and another dies later, so births, deaths and updates all occur.
        let mut ids: Vec<i64> = (0..24).collect();
        if i >= 5 {
            ids.extend(24..28); // born at sample 5
        }
        if i >= 11 {
            ids.retain(|id| *id >= 4); // ids 0..4 die at sample 11
        }
        // Rotate the order so identity, not row position, decides correspondence.
        let shift = i % ids.len().max(1);
        ids.rotate_left(shift);

        let n = ids.len();
        let mut g = GaussianSet {
            positions: Vec::with_capacity(n * 3),
            scales: Vec::with_capacity(n * 3),
            rotations: Vec::with_capacity(n * 4),
            colors: Vec::with_capacity(n * 4),
            motions: Vec::with_capacity(n * 3),
            mu_t: Vec::with_capacity(n),
            sigma_t: Vec::with_capacity(n),
            // One validity window covering the whole clip, stated per gaussian rather than
            // left empty. Empty is not the same claim: the writer would iterate zero windows,
            // fall back to synthesizing `(0, duration)` for the table, and give every gaussian
            // row 0 through `window_index_of`'s `unwrap_or(0)` miss. The file comes out the
            // same as this one, so the binding runners — which do state a window, because the
            // C ABI has no way not to — matched by coincidence rather than by agreement, and
            // any change to the fixture's window would have diverged the Window Table while
            // the gate blamed the binding. Both sides now say the same thing outright.
            win_lo: vec![0.0; n],
            win_hi: vec![DURATION as f32; n],
            ..Default::default()
        };
        for &id in &ids {
            let mut rng = Lcg::seeded(id as u64);
            // A per-gaussian base plus a small per-sample drift: the drift makes position
            // move sample to sample, which is what a non-keyframe sample records as an update.
            let base = [
                rng.unit() * 4.0 - 2.0,
                rng.unit() * 4.0 - 2.0,
                rng.unit() * 4.0 - 2.0,
            ];
            let drift = [
                (rng.unit() - 0.5) * 0.2,
                (rng.unit() - 0.5) * 0.2,
                (rng.unit() - 0.5) * 0.2,
            ];
            for axis in 0..3 {
                g.positions
                    .push((base[axis] + drift[axis] * i as f64) as f32);
                g.scales.push((0.02 + rng.unit() * 0.1) as f32);
            }
            // A unit quaternion, restated absolutely by every update.
            let (qx, qy, qz) = (rng.unit() - 0.5, rng.unit() - 0.5, rng.unit() - 0.5);
            let qw = 1.0;
            let norm = (qx * qx + qy * qy + qz * qz + qw * qw).sqrt();
            g.rotations.extend_from_slice(&[
                (qx / norm) as f32,
                (qy / norm) as f32,
                (qz / norm) as f32,
                (qw / norm) as f32,
            ]);
            g.colors.extend_from_slice(&[
                rng.unit() as f32,
                rng.unit() as f32,
                rng.unit() as f32,
                (0.3 + 0.6 * rng.unit()) as f32,
            ]);
            // A constant velocity per gaussian (a GOP-invariant grid still applies, so the
            // motion bins differ from zero and telescope through the chain).
            g.motions.extend_from_slice(&[
                ((rng.unit() - 0.5) * 0.5) as f32,
                ((rng.unit() - 0.5) * 0.5) as f32,
                ((rng.unit() - 0.5) * 0.5) as f32,
            ]);
            g.mu_t.push((rng.unit() * DURATION) as f32);
            g.sigma_t.push((0.2 + rng.unit() * 0.8) as f32);
        }
        samples.push(Sample {
            t0,
            ids,
            gaussians: g,
        });
    }
    samples
}

/// A tiny linear congruential generator: deterministic, portable, and enough variety for a
/// synthetic scene. Not for anything that needs statistical quality.
struct Lcg {
    state: u64,
}

impl Lcg {
    fn seeded(seed: u64) -> Lcg {
        // Offset so id 0 is not a degenerate all-zero stream.
        Lcg {
            state: seed.wrapping_mul(0x9E37_79B9_7F4A_7C15).wrapping_add(1),
        }
    }

    fn next_u64(&mut self) -> u64 {
        self.state = self
            .state
            .wrapping_mul(6364136223846793005)
            .wrapping_add(1442695040888963407);
        self.state
    }

    /// A value in `[0, 1)`.
    fn unit(&mut self) -> f64 {
        (self.next_u64() >> 11) as f64 / (1u64 << 53) as f64
    }
}

File: cpp/conformance/encode_keyframe_delta.cpp
// Copyright 2026 Avala AI
// SPDX-License-Identifier: Apache-2.0

/// Conformance runner: encode a `keyframe-delta` file (spec §11).
///
/// There is no corpus variant to re-encode here the way `encode_roundtrip` re-encodes one:
/// the keyframe-delta corpus is a whole-file format this package reads through the core's
/// byte-in / string-out surface rather than through an opened `Scene`, so there is no
/// decoded sample sequence to hand back to a writer. This synthesizes one instead —
/// deterministically, from an LCG, so the encoder has no hidden state — matching
/// `rust/conformance/src/bin/encode_keyframe_delta.rs` sample for sample in shape: gaussians
/// move, so every non-keyframe sample carries updates; a band of ids is born partway through
/// and another dies later, so births and deaths both occur; ids are rotated within each
/// sample, so a writer that paired gaussians by row rather than by `gaussian_id` would be
/// wrong; and the cadence lays down more than one keyframe with deltas between them.
///
/// Beside each file it writes the samples it was written from, as JSON. That is what lets
/// the gate around this (`cpp/keyframe-delta-roundtrip.sh`) make the claim no amount of
/// decoder agreement can make: the file, read back by the *Python reference*, against the
/// population that went in and the bounds the file itself declares. An encoder that displaced
/// every position produces a file every decoder reads the same way, and it is wrong.
///
/// Usage: encode_keyframe_delta <out-dir> [chained|keyframe|cadence-one]

#include <algorithm>
#include <cmath>
#include <cstddef>
#include <cstdint>
#include <cstdio>
#include <string>
#include <vector>

#include "fourdgs/fourdgs.hpp"

namespace {

using fourdgs::ErrorCode;
using fourdgs::GaussianData;
using fourdgs::GaussianView;
using fourdgs::KeyframeDeltaOptions;
using fourdgs::KeyframeDeltaSample;
using fourdgs::Result;

constexpr double kDuration = 2.0;
constexpr std::size_t kSamples = 17;

int fail(const std::string& message) {
  std::fprintf(stderr, "%s\n", message.c_str());
  return 1;
}

/// A tiny linear congruential generator: deterministic, portable, and enough variety for a
/// synthetic scene. Not for anything that needs statistical quality. Unsigned overflow wraps
/// by definition in C++, which is what makes this the same generator in every language that
/// has it.
class Lcg {
 public:
  explicit Lcg(std::uint64_t seed)
      // Offset so id 0 is not a degenerate all-zero stream.
      : state_(seed * 0x9E3779B97F4A7C15ull + 1ull) {}

  /// A value in `[0, 1)`.
  double unit() {
    state_ = state_ * 6364136223846793005ull + 1442695040888963407ull;
    return static_cast<double>(state_ >> 11) / static_cast<double>(1ull << 53);
  }

 private:
  std::uint64_t state_;
};

/// One sample's identities and its population.
struct Population {
  std::vector<std::uint32_t> ids;
  GaussianData gaussians;
};

/// Which ids are live at sample `i`, in the order this sample states them.
std::vector<std::uint32_t> idsAt(std::size_t i) {
  std::vector<std::uint32_t> ids;
  for (std::uint32_t id = 0; id < 24; ++id) ids.push_back(id);
  if (i >= 5) {
    for (std::uint32_t id = 24; id < 28; ++id) ids.push_back(id);  // born at sample 5
  }
  if (i >= 11) {
    std::vector<std::uint32_t> survivors;
    for (std::uint32_t id : ids) {
      if (id >= 4) survivors.push_back(id);  // ids 0..3 die at sample 11
    }
    ids = survivors;
  }
  // Rotate, so that identity and not row position decides correspondence.
  const std::size_t shift = ids.empty() ? 0 : i % ids.size();
  std::rotate(ids.begin(), ids.begin() + static_cast<std::ptrdiff_t>(shift), ids.end());
  return ids;
}

Population populationAt(std::size_t i) {
  Population population;
  population.ids = idsAt(i);
  const std::size_t n = population.ids.size();
  GaussianData& g = population.gaussians;
  g.resize(n, 0, 0);
  for (std::size_t row = 0; row < n; ++row) {
    Lcg rng(population.ids[row]);
    // A per-gaussian base plus a small per-sample drift: the drift is what a non-keyframe
    // sample records as an update.
    const double base[3] = {rng.unit() * 4.0 - 2.0, rng.unit() * 4.0 - 2.0, rng.unit() * 4.0 - 2.0};
    const double drift[3] = {(rng.unit() - 0.5) * 0.2, (rng.unit() - 0.5) * 0.2,
                             (rng.unit() - 0.5) * 0.2};
    for (std::size_t axis = 0; axis < 3; ++axis) {
      g.positions[row * 3 + axis] =
          static_cast<float>(base[axis] + drift[axis] * static_cast<double>(i));
      g.scales[row * 3 + axis] = static_cast<float>(0.02 + rng.unit() * 0.1);
    }
    // A unit quaternion, restated absolutely by every update (§11.5).
    const double qx = rng.unit() - 0.5;
    const double qy = rng.unit() - 0.5;
    const double qz = rng.unit() - 0.5;
    const double qw = 1.0;
    const double norm = std::sqrt(qx * qx + qy * qy + qz * qz + qw * qw);
    g.rotations[row * 4 + 0] = static_cast<float>(qx / norm);
    g.rotations[row * 4 + 1] = static_cast<float>(qy / norm);
    g.rotations[row * 4 + 2] = static_cast<float>(qz / norm);
    g.rotations[row * 4 + 3] = static_cast<float>(qw / norm);
    g.colors[row * 4 + 0] = static_cast<float>(rng.unit());
    g.colors[row * 4 + 1] = static_cast<float>(rng.unit());
    g.colors[row * 4 + 2] = static_cast<float>(rng.unit());
    g.colors[row * 4 + 3] = static_cast<float>(0.3 + 0.6 * rng.unit());
    // A constant velocity per gaussian: the motion bins differ from zero and telescope
    // through the chain, which is the property §11.7 is about.
    for (std::size_t axis = 0; axis < 3; ++axis) {
      g.motions[row * 3 + axis] = static_cast<float>((rng.unit() - 0.5) * 0.5);
    }
    g.muT[row] = static_cast<float>(rng.unit() * kDuration);
    g.sigmaT[row] = static_cast<float>(0.2 + rng.unit() * 0.8);
    // One validity window for the whole sequence. `window_index` is GOP-invariant (§11.5),
    // so a fixture that varied it would be exercising the refusal rather than the encode.
    g.winLo[row] = 0.0f;
    g.winHi[row] = static_cast<float>(kDuration);
  }
  return population;
}

/// A number as JSON, at the precision that round-trips a `double` exactly.
///
/// Seventeen significant digits rather than the nine a `float` needs: `t0` is a double, and
/// the gate compares it against the chunk interval the file carries, which is also a double.
/// Nine digits there turns an exact comparison into a near-miss.
std::string number(double value) {
  char buffer[64];
  std::snprintf(buffer, sizeof(buffer), "%.17g", value);
  return buffer;
}

std::string floatArray(const std::vector<float>& values) {
  std::string out = "[";
  for (std::size_t i = 0; i < values.size(); ++i) {
    if (i != 0) out += ",";
    out += number(static_cast<double>(values[i]));
  }
  return out + "]";
}

/// The samples, as the gate reads them back. Hand-written rather than through a JSON library
/// for the reason `check.hpp` has no test framework: a dependency for four object shapes.
Result<void> writeSamplesJson(const std::string& path, const std::vector<Population>& populations,
                              const std::vector<double>& t0s) {
  std::FILE* handle = std::fopen(path.c_str(), "wb");
  if (handle == nullptr) return fourdgs::Error(ErrorCode::kIo, "cannot open " + path);
  std::string out = "{\"durationSec\":" + number(kDuration) + ",\"samples\":[";
  for (std::size_t i = 0; i < populations.size(); ++i) {
    const GaussianData& g = populations[i].gaussians;
    if (i != 0) out += ",";
    out += "{\"t0\":" + number(t0s[i]) + ",\"count\":" + std::to_string(g.count) + ",\"ids\":[";
    for (std::size_t row = 0; row < populations[i].ids.size(); ++row) {
      if (row != 0) out += ",";
      out += std::to_string(populations[i].ids[row]);
    }
    out += "],\"positions\":" + floatArray(g.positions);
    out += ",\"scales\":" + floatArray(g.scales);
    out += ",\"rotations\":" + floatArray(g.rotations);
    out += ",\"colors\":" + floatArray(g.colors);
    out += ",\"motions\":" + floatArray(g.motions);
    out += ",\"muT\":" + floatArray(g.muT);
    out += ",\"sigmaT\":" + floatArray(g.sigmaT);
    out += ",\"winLo\":" + floatArray(g.winLo);
    out += ",\"winHi\":" + floatArray(g.winHi) + "}";
  }
  out += "]}\n";
  std::fwrite(out.data(), 1, out.size(), handle);
  std::fclose(handle);
  return Result<void>();
}

Result<void> writeWhole(const std::string& path, const std::vector<std::uint8_t>& bytes) {
  std::FILE* handle = std::fopen(path.c_str(), "wb");
  if (handle == nullptr) return fourdgs::Error(ErrorCode::kIo, "cannot open " + path);
  if (!bytes.empty()) std::fwrite(bytes.data(), 1, bytes.size(), handle);
  std::fclose(handle);
  return Result<void>();
}

}  // namespace

int main(int argc, char** argv) {
  if (argc < 2 || argc > 3) {
    std::fprintf(stderr, "usage: encode_keyframe_delta <out-dir> [chained|keyframe|cadence-one]\n");
    return 2;
  }
  const std::string directory = argv[1];
  const std::string shape = argc == 3 ? argv[2] : "chained";

  KeyframeDeltaOptions options;
  if (shape == "chained") {
    options.keyframeEvery = 8;
    options.deltaMode = fourdgs::kDeltaModeChained;
  } else if (shape == "keyframe") {
    options.keyframeEvery = 8;
    options.deltaMode = fourdgs::kDeltaModeKeyframe;
  } else if (shape == "cadence-one") {
    // §11.11: every chunk a keyframe. The shape that subsumes the reserved
    // `frame-sequence` name, and the one a writer reaches by cadence alone.
    options.keyframeEvery = 1;
  } else {
    std::fprintf(stderr, "unknown shape %s; expected chained, keyframe or cadence-one\n",
                 shape.c_str());
    return 2;
  }

  std::vector<Population> populations;
  std::vector<double> t0s;
  populations.reserve(kSamples);
  for (std::size_t i = 0; i < kSamples; ++i) {
    populations.push_back(populationAt(i));
    // The t0s tile [0, duration): the last sample starts below the end and its interval
    // closes at `duration`, so no chunk gets a zero-width span (§11.1).
    t0s.push_back(static_cast<double>(i) / static_cast<double>(kSamples) * kDuration);
  }

  // Built after every population is in place: a `GaussianView` borrows, so a view taken from
  // a vector that later reallocates would dangle.
  std::vector<KeyframeDeltaSample> samples;
  samples.reserve(kSamples);
  for (std::size_t i = 0; i < kSamples; ++i) {
    KeyframeDeltaSample sample;
    sample.t0 = t0s[i];
    sample.ids =
        fourdgs::Span<const std::uint32_t>(populations[i].ids.data(), populations[i].ids.size());
    sample.gaussians = GaussianView(populations[i].gaussians);
    samples.push_back(sample);
  }
  const fourdgs::Span<const KeyframeDeltaSample> span(samples.data(), samples.size());

  Result<std::vector<std::uint8_t>> encoded =
      fourdgs::encodeKeyframeDeltaSequence(span, kDuration, options);
  if (!encoded) return fail(encoded.error().toString());

  // Two encodes of one sequence must be the same bytes. A writer that iterated a map, or
  // sorted unstably, passes every value-based check and still produces a file that differs
  // between runs — a property nobody notices until a build is expected to reproduce.
  Result<std::vector<std::uint8_t>> again =
      fourdgs::encodeKeyframeDeltaSequence(span, kDuration, options);
  if (!again) return fail(again.error().toString());
  if (*again != *encoded) {
    return fail(shape + ": two encodes of one sequence differ; the writer is not deterministic");
  }

  // The file has to survive both of this package's read paths, and they must reach the same
  // populations, before it is worth asking another implementation about it.
  const fourdgs::Span<const std::uint8_t> bytes(encoded->data(), encoded->size());
  Result<std::string> streamed = fourdgs::keyframeDeltaStatesJson(bytes, false);
  if (!streamed)
    return fail(shape + ": the writer wrote a file it cannot stream-decode: " +
                streamed.error().toString());
  Result<std::string> indexed = fourdgs::keyframeDeltaStatesJson(bytes, true);
  if (!indexed)
    return fail(shape +
                ": the writer wrote a file it cannot index-decode: " + indexed.error().toString());
  if (*streamed != *indexed) {
    return fail(shape + ": the two read paths disagree on a file the writer produced");
  }

  const std::string stem = directory + "/keyframe-delta-" + shape;
  Result<void> written = writeWhole(stem + ".4dgs", *encoded);
  if (!written) return fail(written.error().toString());
  written = writeSamplesJson(stem + ".samples.json", populations, t0s);
  if (!written) return fail(written.error().toString());

  std::printf("keyframe-delta-%s %zu samples, %zu bytes, deterministic, both read paths agree\n",
              shape.c_str(), samples.size(), encoded->size());
  return 0;
}
